### PR TITLE
Support type relaxation for JSON import

### PR DESCRIPTION
**Pull request by dominiklohmann**

### CODE REVIEW

File: libvast/vast/concept/parseable/vast/json.hpp [hunk 1]
```diff
@@ -13,17 +13,30 @@
...
+namespace parsers {
+
+static auto const json_boolean = ignore(*parsers::space) >> parsers::boolean;
+static auto const json_number
+  = ignore(*parsers::space) >> parsers::real_opt_dot;
+
+// These parsers are only needed for relaxed conversion from JSON string to VAST
+// type, they are not part of the actual JSON specification.
+static auto const json_int = ignore(*parsers::space) >> parsers::i64;
+static auto const json_count = ignore(*parsers::space) >> parsers::u64;
```
Comment (mavam) [inline libvast/vast/concept/parseable/vast/json.hpp:36]:
```suggestion
static auto const json_boolean = ignore(*space) >> boolean;
static auto const json_number = ignore(*space) >> real_opt_dot;

// These parsers are only needed for relaxed conversion from JSON string to VAST
// type, they are not part of the actual JSON specification.
static auto const json_int = ignore(*space) >> i64;
static auto const json_count = ignore(*space) >> u64;
```

File: libvast/src/format/json.cpp [hunk 2]
```diff
@@ -145,6 +145,43 @@ struct convert {
   }
 };
 
+struct relaxed_convert : strict_convert {
+  using strict_convert::operator();
+
+  caf::expected<data> operator()(json::string str, const bool_type&) const {
```
Comment (mavam) [inline libvast/src/format/json.cpp:151]:
```suggestion
  caf::expected<data> operator()(const json::string& str, const bool_type&) const {
```

File: libvast/src/format/json.cpp [hunk 3]
```diff
@@ -145,6 +145,43 @@ struct convert {
...
+    if (bool res; parsers::json_boolean(str, res))
+      return res;
+    return make_error(ec::convert_error, "cannot convert from", str, "to bool");
+  }
+
+  caf::expected<data> operator()(json::string str, const real_type&) const {
```
Comment (mavam) [inline libvast/src/format/json.cpp:157]:
```suggestion
  caf::expected<data> operator()(const json::string& str, const real_type&) const {
```

File: libvast/src/format/json.cpp [hunk 4]
```diff
@@ -145,6 +145,43 @@ struct convert {
...
+    if (real res; parsers::json_number(str, res))
+      return res;
+    return make_error(ec::convert_error, "cannot convert from", str, "to real");
+  }
+
+  caf::expected<data> operator()(json::string str, const integer_type&) const {
```
Comment (mavam) [inline libvast/src/format/json.cpp:163]:
```suggestion
  caf::expected<data> operator()(const json::string& str, const integer_type&) const {
```

File: libvast/src/format/json.cpp [hunk 5]
```diff
@@ -145,6 +145,43 @@ struct convert {
...
+    if (integer res; parsers::json_int(str, res))
+      return res;
+    if (real res; parsers::json_number(str, res)) {
+      VAST_WARNING_ANON("json-reader narrowed", str, "to type int");
+      return detail::narrow_cast<integer>(res);
+    }
+    return make_error(ec::convert_error, "cannot convert from", str, "to int");
+  }
+
+  caf::expected<data> operator()(json::string str, const count_type&) const {
```
Comment (mavam) [inline libvast/src/format/json.cpp:173]:
```suggestion
  caf::expected<data> operator()(const json::string& str, const count_type&) const {
```

File: libvast/src/format/json.cpp [hunk 6]
```diff
@@ -145,6 +145,43 @@ struct convert {
   }
 };
 
+struct relaxed_convert : strict_convert {
+  using strict_convert::operator();
+
+  caf::expected<data> operator()(json::string str, const bool_type&) const {
+    if (bool res; parsers::json_boolean(str, res))
```
Comment (mavam) [inline libvast/src/format/json.cpp:152]:
```suggestion
    if (bool x; parsers::json_boolean(str, res))
```

I think just using `x` would make things a bit more readable. The `res` things feel like a sub-optimal compromise between `result` and a less verbose token.

File: libvast/src/format/json.cpp [hunk 7]
```diff
@@ -189,7 +226,14 @@ caf::error add(table_slice_builder& builder, const vast::json::object& xs,
                                            "slice builder");
       continue;
     }
-    auto x = caf::visit(convert{}, *i, field.type);
+    auto x = [&] {
+      switch (policy) {
```
Comment (mavam) [inline libvast/src/format/json.cpp:230]: Can you add a `default` case with a failing assertion?

Comment (dominiklohmann) [reply libvast/src/format/json.cpp:230]: It's an enum class and `-Wswitch` is enabled by default. Adding a default case here is an antipattern as it silences compiler warnings.

**Comment by dominiklohmann** [comment]
I rebased this onto current master to solve the conflict.

File: libvast/vast/concept/parseable/numeric/integral.hpp [hunk 8]
```diff
@@ -32,32 +33,44 @@ bool parse_sign(Iterator& i) {
...
   using attribute = T;
 
   static bool isdigit(char c) {
-    return c >= '0' && c <= '9';
+    if constexpr (Radix == 10)
+      return c >= '0' && c <= '9';
+    else if constexpr (Radix == 16)
+      return std::isxdigit(c);
+    else
+      static_assert(detail::always_false_v<decltype(Radix)>, "unsupported "
```
Comment (mavam) [inline libvast/vast/concept/parseable/numeric/integral.hpp:56]: Why do you need `decltype` here?

Comment (dominiklohmann) [reply libvast/vast/concept/parseable/numeric/integral.hpp:56]: Radix is an integer value, not a type, and the latter is expected by the type trait `detail::always_false`.

File: libvast/vast/concept/parseable/vast/json.hpp [hunk 9]
```diff
@@ -13,17 +13,32 @@
...
+namespace parsers {
+
+static auto const json_boolean = parsers::boolean;
+static auto const json_int = parsers::i64;
+static auto const json_count = (!parsers::hex_prefix >> parsers::u64)
+                               | (parsers::hex_prefix >> parsers::hex64);
+static auto const json_number
+  = (!parsers::hex_prefix >> parsers::real_opt_dot)
+    | (parsers::hex_prefix >> parsers::hex64->*[](uint64_t x) {
+        return detail::narrow_cast<double>(x);
```
Comment (mavam) [inline libvast/vast/concept/parseable/vast/json.hpp:37]:
```suggestion
        return detail::narrow_cast<json::number>(x);
```

File: libvast/vast/concept/parseable/vast/json.hpp [hunk 10]
```diff
@@ -13,17 +13,32 @@
...
 
+namespace parsers {
+
+static auto const json_boolean = parsers::boolean;
+static auto const json_int = parsers::i64;
+static auto const json_count = (!parsers::hex_prefix >> parsers::u64)
+                               | (parsers::hex_prefix >> parsers::hex64);
+static auto const json_number
+  = (!parsers::hex_prefix >> parsers::real_opt_dot)
+    | (parsers::hex_prefix >> parsers::hex64->*[](uint64_t x) {
```
Comment (mavam) [inline libvast/vast/concept/parseable/vast/json.hpp:36]: Why didn't you attempt to parse as hex first? Like this:

```cpp
  (parsers::hex_prefix >> parsers::hex64)
| parsers::real_opt_dot
```

File: libvast/vast/concept/parseable/vast/json.hpp [hunk 11]
```diff
@@ -39,15 +54,20 @@ struct json_parser : parser<json_parser> {
...
     auto object = as<json::object>(lbrace >> ~(key_value % delim) >> rbrace);
-    j = null | boolean | number | string | array | object;
+    // clang-format off
+    j = ws >> (null
+            | json_boolean
+            | json_number
+            | string
+            | array
+            | object);
+    // clang-format on
```
Comment (mavam) [inline libvast/vast/concept/parseable/vast/json.hpp:70]:
```suggestion
    j = ws >> ( null
              | json_boolean
              | json_number
              | string
              | array
              | object
              );
    // clang-format on
```

File: libvast/test/parseable.cpp [hunk 12]
```diff
@@ -501,6 +501,24 @@ TEST(unsigned integral) {
...
+  CHECK(p("13BFC3d1", x));
+  CHECK_EQUAL(x, 0x13BFC3d1);
+  CHECK(p("FF", x));
+  CHECK_EQUAL(x, 0xFF);
+  CHECK(p("ff00", x));
+  CHECK_EQUAL(x, 0xff00);
+  CHECK(p("0X12ab", x));
+  CHECK_EQUAL(x, 0X12ab);
+  CHECK(p("0x3e7", x));
+  CHECK_EQUAL(x, 0x3e7);
```
Comment (mavam) [inline libvast/test/parseable.cpp:519]: What about things like `0x0000aa`?

File: vast.conf [hunk 13]
```diff
@@ -77,7 +77,10 @@ export {
   ; The `vast export json` command exports events formatted as JSONL (line-
   ; delimited JSON).
   json {
-    ; For available options, see export.ascii.
+    ; Strict type matching.
+    ;strict = false
+
+    ; For additionally available options, see export.ascii.
```
Comment (mavam) [inline vast.conf:83]: This needs to be removed as well.

File: libvast/vast/format/json.hpp [hunk 14]
```diff
@@ -153,7 +154,7 @@ class reader final : public multi_layout_reader {
 
 template <class Selector>
 reader<Selector>::reader(caf::atom_value table_slice_type,
-                         const caf::settings& /*options*/,
+                         const caf::settings& options,
```
Comment (mavam) [inline libvast/vast/format/json.hpp:157]: This can be commented again.

File: libvast/vast/format/json.hpp [hunk 15]
```diff
@@ -31,6 +31,7 @@
 
 #include <caf/expected.hpp>
 #include <caf/fwd.hpp>
+#include <caf/settings.hpp>
```
Comment (mavam) [inline libvast/vast/format/json.hpp:34]: Not needed anymore.

**Comment by dominiklohmann** [comment]
Done @mavam.

**Pull request merged**